### PR TITLE
Add Resource Server Sample

### LINKED ISSUE #4: Add Resource Server Sample
Related to

It would be good to have a Resource Server sample that integrates with the authorization server sample. It should be quite similar to [the Resource Server sample](https://github.com/spring-projects/spring-security/tree/master/samples/boot/oauth2resourceserver/src) in Spring Security.

Like other samples, it should go into the `samples` directory. For simplicity, classes ought to be placed into a package called `sample`.

Given a valid token from the authorization server, the following test should pass:

```java
this.mockMvc.perform(get("/")
        .header("Authorization", "Bearer " + token))
        .andExpect(status().isOk());
```

### PR DESCRIPTION
**Pull request by ketola**
Add Resource Server Sample
Fixes #4 

- I configured the resource server based on the instructions at https://docs.spring.io/spring-security/site/docs/current/reference/html5/#oauth2resourceserver
- The token validation obviously fails until the auth server is able to provide the token. And as an example, I left the issuer uri to point at https://accounts.google.com.
- I noticed there was a separate issue about configuring the gradle build, that's why I left the module as a standalone.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@ketola Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=30)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=30) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@ketola Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=30)!

File: samples/resource-server-sample/bin/main/application.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# See https://docs.spring.io/spring-security/site/docs/current/reference/html5/#oauth2resourceserver
```
Comment (anzap) [inline samples/resource-server-sample/bin/main/application.yaml:1]: Files in `/bin` should not be committed. PR #31 adds this folder to .gitignore.

File: samples/resource-server-sample/build.gradle [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'org.springframework.security:spring-security-oauth2-resource-server:5.3.1.RELEASE'
+	implementation 'org.springframework.security:spring-security-oauth2-jose:5.3.1.RELEASE'
+	implementation 'org.springframework.security:spring-security-config:5.3.1.RELEASE'
+	
+	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+}
+
+test {
+	useJUnitPlatform()
```
Comment (anzap) [inline samples/resource-server-sample/build.gradle:25]: Should we use JUnit4 or JUnit5 for testing?

Comment (ketola) [reply samples/resource-server-sample/build.gradle:27]: I changed it to use junit 5 api, as that was also used in the parent

**Comment by ketola** [comment]
> I see you added this under `samples` folder. Project does not seem to have gone the multimodule way as there is now a top-level src folder with a `sample` package. Not sure this should be cleaned up now or will be part of issue #10 .

I left the module as standalone on purpose as I saw that there was another issue for creating the gradle build.

File: samples/resource-server-sample/build.gradle [hunk 3]
```diff
@@ -0,0 +1,28 @@
+plugins {
```
Comment (jgrandja) [inline samples/resource-server-sample/build.gradle:1]: Please modify the build to follow the [helloworld](https://github.com/spring-projects-experimental/spring-authorization-server/blob/master/samples/helloworld/spring-authorization-server-samples-helloworld.gradle) sample, which uses the [spring-build-conventions](https://github.com/spring-gradle-plugins/spring-build-conventions) plugin (same as Spring Security).

The build file should be renamed to `spring-authorization-server-samples-oauth2resourceserver.gradle`

File: samples/resource-server-sample/src/main/java/sample/SampleResource.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sample;
+
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class SampleResource {
```
Comment (jgrandja) [inline samples/resource-server-sample/src/main/java/sample/SampleResource.java:22]: Can we rename this to something more meaningful? Maybe `ResourceController` and `resource()`?

File: samples/resource-server-sample/src/main/resources/application.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# See https://docs.spring.io/spring-security/site/docs/current/reference/html5/#oauth2resourceserver 
+# for more configuration options
+spring:
+  security:
+    oauth2:
+      resourceserver:
+        jwt:
+          issuer-uri:  https://accounts.google.com
```
Comment (jgrandja) [inline samples/resource-server-sample/src/main/resources/application.yaml:8]: Let's use the `jwk-set-uri` property instead, that way the resource server will at least startup.

We could also mock the jwkSet endpoint using `MockWebServer`. Take a look at `OAuth2ResourceServerConfigurerTests` in Spring Security for an example on how to set it up.

File: samples/resource-server-sample/build.gradle (hunk 3 above)
Comment (ketola) [reply samples/resource-server-sample/build.gradle]: Done.

File: samples/resource-server-sample/src/main/java/sample/SampleResource.java (hunk 4 above)
Comment (ketola) [reply samples/resource-server-sample/src/main/java/sample/SampleResource.java]: Done.

File: samples/resource-server-sample/src/main/resources/application.yaml (hunk 5 above)
Comment (ketola) [reply samples/resource-server-sample/src/main/resources/application.yaml]: With this one, I will need some more time as I have not used the `MockWebServer` before, I'll do some research, check the example and try to finish this later today.

**Comment by ketola** [comment]
@jgrandja, I noticed you labeled this as 'duplicate', where has this work been duplicated?

**Comment by jgrandja** [comment]
@ketola We mark PR's as duplicate if there is an associated issue, which there is in this case. However, we typically mark it when we merge. I'll remove the label until we merge to avoid confusion.

File: samples/resource-server-sample/src/main/resources/application.yaml (hunk 5 above)
Comment (ketola) [reply samples/resource-server-sample/src/main/resources/application.yaml]: Ok, I checked out the sample from OAuth2ResourceServerConfigurerTests with the MockWebServer, but It started to feel like a lot's work to get the token validation mocked that way. Instead, I decided to mock the JwtDecoder. Let me know what you think.

**Comment by jgrandja** [comment]
Thanks for the updates @ketola. This is now in master.

**Pull request closed without merging**